src/bin/basic_subclass.rs

```diff
@@ -0,0 +1,179 @@
+//! # Basic Subclass example
+
+#[macro_use]
+extern crate glib;
+extern crate gio;
+extern crate gtk;
+
+use gio::prelude::*;
+use gtk::prelude::*;
+
+use glib::subclass;
+use glib::subclass::prelude::*;
+use glib::translate::*;
+use gtk::subclass::prelude::*;
+
+use gio::ApplicationFlags;
+use std::cell::RefCell;
+
+#[derive(Debug)]
+struct WindowState {
+    headerbar: gtk::HeaderBar,
+    label: gtk::Label,
+}
+
+#[derive(Debug)]
+pub struct SimpleAppWindowPrivate {
+    state: RefCell<Option<WindowState>>,
+}
+
+impl ObjectSubclass for SimpleAppWindowPrivate {
+    const NAME: &'static str = "SimpleAppWindowPrivate";
+    type ParentType = gtk::ApplicationWindow;
+    type Instance = subclass::simple::InstanceStruct<Self>;
+    type Class = subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn new() -> Self {
+        Self {
+            state: RefCell::new(None),
+        }
+    }
+}
+
+impl ObjectImpl for SimpleAppWindowPrivate {
+    glib_object_impl!();
+
+    fn constructed(&self, obj: &glib::Object) {
+        self.parent_constructed(obj);
+        let self_ = obj.downcast_ref::<SimpleAppWindow>().unwrap();
+
+        let headerbar = gtk::HeaderBar::new();
+        headerbar.set_title(Some("Hello World!"));
+        headerbar.set_show_close_button(true);
+
+        let label = gtk::Label::new(Some("Hello World!"));
+
+        self_.add(&label);
+        self_.set_titlebar(Some(&headerbar));
+        self_.set_default_size(640, 480);
+
+        *self.state.borrow_mut() = Some(WindowState { headerbar, label });
+    }
+}
+
+impl WidgetImpl for SimpleAppWindowPrivate {}
+impl ContainerImpl for SimpleAppWindowPrivate {}
+impl BinImpl for SimpleAppWindowPrivate {}
+impl WindowImpl for SimpleAppWindowPrivate {}
+impl ApplicationWindowImpl for SimpleAppWindowPrivate {}
+
+glib_wrapper! {
+    pub struct SimpleAppWindow(
+        Object<subclass::simple::InstanceStruct<SimpleAppWindowPrivate>,
+        subclass::simple::ClassStruct<SimpleAppWindowPrivate>,
+        SimpleAppWindowClass>)
+        @extends gtk::Widget, gtk::Container, gtk::Bin, gtk::Window, gtk::ApplicationWindow;
+
+    match fn {
+        get_type => || SimpleAppWindowPrivate::get_type().to_glib(),
+    }
+}
+
+impl SimpleAppWindow {
+    pub fn new(app: &gtk::Application) -> Self {
+        glib::Object::new(Self::static_type(), &[("application", app)])
+            .expect("Failed to create SimpleAppWindow")
+            .downcast::<SimpleAppWindow>()
+            .expect("Created simpleappwindow is of wrong type")
+    }
+}
+
+#[derive(Debug)]
+struct State {
```

Comment:
It's a bit awkward to have these separate structs. Maybe remove them?

Reply:
They are, but its also ugly having a `RefCell<Option<T>>` for every widget field.

Reply:
Maybe instead of State, call these structs Widgets?

Reply:
Let's call the one in the app `window` and directly store the window, and the other struct let's call it `Widgets` and store all the widgets in there.